File: common/src/main/java/net/impleri/playerskills/utils/PlayerSkillsLogger.java
package net.impleri.playerskills.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;

public class PlayerSkillsLogger {

    public static PlayerSkillsLogger create(String modId) {
        return create(modId, null);
    }

    public static PlayerSkillsLogger create(String modId, @Nullable String prefix) {

        return new PlayerSkillsLogger(modId, prefix);
    }

    private final Logger instance;

    private boolean debug = false;

    private final String prefix;

    private PlayerSkillsLogger(String modId, String prefix) {
        this.prefix = prefix;
        this.instance = LogManager.getLogger(modId);
    }

    public void enableDebug() {
        debug = true;
    }

    public void disableDebug() {
        debug = false;
    }

    public boolean toggleDebug() {
        debug = !debug;

        return debug;
    }

    private String addPrefix(String message) {
        return "[" + prefix + "] " + message;
    }

    public void error(String message) {
        if (instance != null) {
            instance.error(addPrefix(message));
        }
    }

    public void error(String message, Object... params) {
        if (instance != null) {
            instance.error(addPrefix(message), Arrays.stream(params).toArray());
        }
    }

    public void warn(String message) {
        instance.warn(addPrefix(message));
    }

    public void warn(String message, Object... params) {
        instance.warn(addPrefix(message), Arrays.stream(params).toArray());
    }

    public void info(String message) {
        instance.info(addPrefix(message));
    }

    public void info(String message, Object... params) {
        instance.info(addPrefix(message), Arrays.stream(params).toArray());
    }

    public void debug(String message) {
        if (debug) {
            info(message);
            return;
        }
        instance.debug(addPrefix(message));
    }

    public void debug(String message, Object... params) {
        if (debug) {
            info(message, params);
            return;
        }
        instance.debug(addPrefix(message), Arrays.stream(params).toArray());
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/utils/PlayerSkillsLogger.kt
package net.impleri.playerskills.utils

import net.impleri.playerskills.PlayerSkills
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

class PlayerSkillsLogger private constructor(modId: String, private val prefix: String) {
  private val instance: Logger
  private var debug = false

  init {
    instance = LogManager.getLogger(modId)
  }

  fun enableDebug() {
    debug = true
  }

  fun disableDebug() {
    debug = false
  }

  fun toggleDebug(): Boolean {
    debug = !debug
    return debug
  }

  private fun addPrefix(message: String): String {
    return "[$prefix] $message"
  }

  fun error(message: String) {
    instance.error(addPrefix(message))
  }

  fun warn(message: String) {
    instance.warn(addPrefix(message))
  }

  fun info(message: String) {
    instance.info(addPrefix(message))
  }

  fun debug(message: String) {
    if (debug) {
      info(message)
      return
    }
    instance.debug(addPrefix(message))
  }

  companion object {
    fun create(modId: String, prefix: String = "MOD"): PlayerSkillsLogger {
      return PlayerSkillsLogger(modId, prefix)
    }

    fun create(prefix: String = "MOD"): PlayerSkillsLogger {
      return PlayerSkillsLogger(PlayerSkills.MOD_ID, prefix)
    }

    enum class TYPES {
      BLOCKS,
      FLUIDS,
      ITEMS,
      MOBS,
      SKILLS,
      SKIPS,
    }

    fun toggleDebug(type: TYPES?): Boolean {
      return when (type) {
        TYPES.BLOCKS -> BLOCKS.toggleDebug()
        TYPES.FLUIDS -> FLUIDS.toggleDebug()
        TYPES.ITEMS -> ITEMS.toggleDebug()
        TYPES.MOBS -> MOBS.toggleDebug()
        TYPES.SKIPS -> SKIPS.toggleDebug()
        else -> SKILLS.toggleDebug()
      }
    }

    internal val SKILLS = create("SKILLS")

    internal val SKIPS = create("REST")

    internal val BLOCKS = create("BLOCKS")

    internal val FLUIDS = create("FLUIDS")

    internal val ITEMS = create("ITEMS")

    internal val MOBS = create("MOBS")
  }
}
